lib/GHCup/Version.hs

```diff
@@ -65,44 +65,15 @@ pvpToVersion pvp_ rest =
 -- -- prop> \v -> let (Just (pvp', r)) = versionToPVP v in pvpToVersion pvp' r === Just v
 versionToPVP :: MonadThrow m => V.Version -> m (V.PVP, Text)
 versionToPVP (V.Version (Just _) _ _ _) = throwM $ ParseError "Unexpected epoch"
-versionToPVP v = either (\_ -> (, rest v) <$> alternative v) (pure . (, mempty)) . V.pvp . V.prettyVer $ v
+versionToPVP v = case parse pvp'' "Version->PVP" $ V.prettyVer v of
+  Left _  -> throwM $ ParseError "Couldn't convert Version to PVP"
+  Right r -> pure r
  where
-  alternative :: MonadThrow m => V.Version -> m V.PVP
-  alternative v' = case NE.takeWhile isDigit (V._vChunks v') of
-    [] -> throwM $ ParseError "Couldn't convert Version to PVP"
-    xs -> pure $ pvpFromList (unsafeDigit <$> xs)
-
-  rest :: V.Version -> Text
-  rest (V.Version _ cs pr me) =
-    let chunks = NE.dropWhile isDigit cs
-        ver = intersperse (T.pack ".") . chunksAsT $ chunks
-        me' = maybe [] (\m -> [T.pack "+",m]) me
-        pr' = foldable [] (T.pack "-" :) $ intersperse (T.pack ".") (chunksAsT pr)
-        prefix = case (ver, pr', me') of
-                   (_:_, _, _) -> T.pack "."
-                   _           -> T.pack ""
-    in prefix <> mconcat (ver <> pr' <> me')
-   where
-    chunksAsT :: Functor t => t V.VChunk -> t Text
-    chunksAsT = fmap (foldMap f)
-      where
-        f :: V.VUnit -> Text
-        f (V.Digits i) = T.pack $ show i
-        f (V.Str s)    = s
-
-    foldable :: Foldable f => f b -> (f a -> f b) -> f a -> f b
-    foldable d g f | null f    = d
-                   | otherwise = g f
-
-
-
-  isDigit :: V.VChunk -> Bool
-  isDigit (V.Digits _ :| []) = True
-  isDigit _                = False
-
-  unsafeDigit :: V.VChunk -> Int
-  unsafeDigit (V.Digits x :| []) = fromIntegral x
-  unsafeDigit _ = error "unsafeDigit: wrong input"
+   pvp'' :: Parsec Void T.Text (V.PVP, T.Text)
+   pvp'' = do
+     p <- V.pvp'
+     s <- getParserState
+     pure (p, stateInput s)
```

Comment:
Instead of manually traversing into the structure of the `Version`, this just reparses the `PVP` normally and yields whatever was left by using `megaparsec` functionality.